File: src/main/java/br/com/fiap/gestaoestoque/model/Produto.java
package br.com.fiap.gestaoestoque.model;

import br.com.fiap.gestaoestoque.model.enums.CategoriaProduto;
import jakarta.persistence.*;
import jakarta.validation.constraints.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * Entidade representando um Produto no sistema de gestão de estoque.
 * 
 * Regras de negócio:
 * - Código SKU deve ser único
 * - Preço unitário deve ser positivo
 * - Quantidade mínima em estoque não pode ser negativa
 */
@Entity
@Table(name = "produtos")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Produto {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(unique = true, nullable = false, length = 50)
    @NotBlank(message = "Código SKU é obrigatório")
    @Size(max = 50, message = "Código SKU deve ter no máximo 50 caracteres")
    private String codigoSku;

    @Column(nullable = false, length = 200)
    @NotBlank(message = "Nome do produto é obrigatório")
    @Size(max = 200, message = "Nome deve ter no máximo 200 caracteres")
    private String nome;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    @NotNull(message = "Categoria é obrigatória")
    private CategoriaProduto categoria;

    @Column(nullable = false, precision = 10, scale = 2)
    @NotNull(message = "Preço unitário é obrigatório")
    @DecimalMin(value = "0.01", message = "Preço unitário deve ser maior que zero")
    private BigDecimal precoUnitario;

    @Column(nullable = false)
    @NotNull(message = "Quantidade mínima em estoque é obrigatória")
    @Min(value = 0, message = "Quantidade mínima não pode ser negativa")
    private Integer quantidadeMinima;

    @Column(nullable = false)
    private Integer quantidadeAtual = 0;

    @Column(nullable = false, updatable = false)
    private LocalDateTime dataCriacao;

    @OneToMany(mappedBy = "produto", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<MovimentacaoEstoque> movimentacoes = new ArrayList<>();

    @PrePersist
    protected void onCreate() {
        dataCriacao = LocalDateTime.now();
        if (quantidadeAtual == null) {
            quantidadeAtual = 0;
        }
    }

    /**
     * Verifica se o produto está abaixo da quantidade mínima (alerta de estoque baixo).
     */
    public boolean isEstoqueBaixo() {
        return quantidadeAtual < quantidadeMinima;
    }

    /**
     * Verifica se é um produto perecível.
     */
    public boolean isPerecivel() {
        return categoria == CategoriaProduto.PERECIVEL;
    }

    /**
     * Adiciona quantidade ao estoque atual.
     */
    public void adicionarEstoque(Integer quantidade) {
        this.quantidadeAtual += quantidade;
    }

    /**
     * Remove quantidade do estoque atual.
     */
    public void removerEstoque(Integer quantidade) {
        this.quantidadeAtual -= quantidade;
    }

    /**
     * Verifica se há estoque suficiente para retirada.
     */
    public boolean temEstoqueSuficiente(Integer quantidade) {
        return this.quantidadeAtual >= quantidade;
    }
}

File: GestaoEstoque/Models/Produto.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using GestaoEstoque.Models.Enums;

namespace GestaoEstoque.Models;

/// <summary>
/// Entidade representando um Produto no sistema de gestão de estoque.
/// 
/// Regras de negócio:
/// - Código SKU deve ser único
/// - Preço unitário deve ser positivo
/// - Quantidade mínima em estoque não pode ser negativa
/// </summary>
[Table("Produtos")]
public class Produto
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public long Id { get; set; }

    [Required(ErrorMessage = "Código SKU é obrigatório")]
    [StringLength(50, ErrorMessage = "Código SKU deve ter no máximo 50 caracteres")]
    public string CodigoSku { get; set; } = string.Empty;

    [Required(ErrorMessage = "Nome do produto é obrigatório")]
    [StringLength(200, ErrorMessage = "Nome deve ter no máximo 200 caracteres")]
    public string Nome { get; set; } = string.Empty;

    [Required(ErrorMessage = "Categoria é obrigatória")]
    public CategoriaProduto Categoria { get; set; }

    [Required(ErrorMessage = "Preço unitário é obrigatório")]
    [Range(0.01, double.MaxValue, ErrorMessage = "Preço unitário deve ser maior que zero")]
    [Column(TypeName = "decimal(10,2)")]
    public decimal PrecoUnitario { get; set; }

    [Required(ErrorMessage = "Quantidade mínima em estoque é obrigatória")]
    [Range(0, int.MaxValue, ErrorMessage = "Quantidade mínima não pode ser negativa")]
    public int QuantidadeMinima { get; set; }

    public int QuantidadeAtual { get; set; } = 0;

    [Required]
    public DateTime DataCriacao { get; set; } = DateTime.Now;

    // Relacionamento com Movimentações
    public virtual ICollection<MovimentacaoEstoque> Movimentacoes { get; set; } = new List<MovimentacaoEstoque>();

    /// <summary>
    /// Verifica se o produto está abaixo da quantidade mínima (alerta de estoque baixo).
    /// </summary>
    public bool IsEstoqueBaixo() => QuantidadeAtual < QuantidadeMinima;

    /// <summary>
    /// Verifica se é um produto perecível.
    /// </summary>
    public bool IsPerecivel() => Categoria == CategoriaProduto.PERECIVEL;

    /// <summary>
    /// Adiciona quantidade ao estoque atual.
    /// </summary>
    public void AdicionarEstoque(int quantidade) => QuantidadeAtual += quantidade;

    /// <summary>
    /// Remove quantidade do estoque atual.
    /// </summary>
    public void RemoverEstoque(int quantidade) => QuantidadeAtual -= quantidade;

    /// <summary>
    /// Verifica se há estoque suficiente para retirada.
    /// </summary>
    public bool TemEstoqueSuficiente(int quantidade) => QuantidadeAtual >= quantidade;
}
